File: ffi/ts/src/keys/doublePublicKey.ts
import {
  castToDpk,
  decodeAddress,
  deserializeDpk,
  encodeAddress,
  freeObj,
  genDoublePubKey,
  genDpkWithKeysAcctAddr,
  serializeDpk,
} from '../blsct'

import { ManagedObj, isWasmPtrWrapper } from '../managedObj'
import { PublicKey } from './publicKey'
import { Scalar } from '../scalar'

// Address encoding types
const AddressEncoding = {
  Bech32: 0,
  Bech32M: 1,
} as const

/** The unique source from which an address is derived.
 * A `DoublePublicKey` is a pair of `PublicKey`s that can be used to derive an address.
 *
 * Instantiating a `DoublePublicKey` object without a parameter returns a `DoublePublicKey` consisting of two randomly generated `PublicKey`s.
 *
 * Examples:
 * ```ts
 * const { DoublePublicKey, PublicKey } = require('navio-blsct')
 * const dpk = new DoublePublicKey()
 * const pk1 = PublicKey.random()
 * const pk2 = PublicKey.random()
 * const dpk2 = DoublePublicKey.from_public_keys(pk1, pk2)
 * const vk = Scalar.random()
 * const spendingPk = PublicKey.random()
 * const dpk3 = DoublePublicKey.fromKeysAndAcctAddr(vk, spendingPk, 1, 2)
 * const ser = dpk3.serialize()
 * const deser = DoublePublicKey.deserialize(ser)
 * deser.serialize() == ser
 * ```
 */ 
export class DoublePublicKey extends ManagedObj {
  constructor(obj?: any) {
    if (isWasmPtrWrapper(obj)) {
      // WASM pointer from fromObj or _deserialize
      super(obj)
    } else if (typeof obj === 'object' && obj !== null) {
      // Native NAPI object
      super(obj)
    } else {
      const pk1 = PublicKey.random()
      const pk2 = PublicKey.random()

      const rv = genDoublePubKey(pk1.value(), pk2.value())
      if (rv.result !== 0) {
        freeObj(rv)
        throw new Error(`Failed to generate DoublePublicKey: ${rv.result}`)
      }
      super(rv.value)
      freeObj(rv)
    }
  }

  // TODO add equals, getPk1 and getPk2 methods in libblsct

  /** Generates a random `DoublePublicKey`.
   * @returns A new DoublePublicKey instance with two random `PublicKey`s.
   */
  static random(): DoublePublicKey {
    const vk = PublicKey.random()
    const sk = PublicKey.random()
    return DoublePublicKey.fromViewAndSpendKeys(vk, sk)
  }

  /** Generates a `DoublePublicKey` from the provided view key and spend keys.
   * @param vk - the view key.
   * @param sk - The spend key.
   * @returns A new `DoublePublicKey` instance.
   */
  static fromViewAndSpendKeys(
    vk: PublicKey,
    sk: PublicKey,
  ): DoublePublicKey {
    const rv = genDoublePubKey(vk.value(), sk.value())
    const dpk = DoublePublicKey.fromObj(rv.value)
    freeObj(rv)
    return dpk
  }

  /** Generates a `DoublePublicKey` from the provided `PublicKey`, account, and address 
   * @param viewKey - The `Scalar` used to derive the `DoublePublicKey`.
   * @param spendingPubKey - The `PublicKey` used for spending.
   * @param account - The account.
   * @param address - The address.
   * @returns A new `DoublePublicKey` instance.
   */
  static fromKeysAcctAddr(
    viewKey: Scalar,
    spendingPubKey: PublicKey,
    account: number,
    address: number,
  ): DoublePublicKey {
    const obj = genDpkWithKeysAcctAddr(
      viewKey.value(),
      spendingPubKey.value(),
      account,
      address
    )
    return DoublePublicKey.fromObj(obj) 
  }

  override value(): any {
    return castToDpk(this.obj)
  }

  override serialize(): string {
    return serializeDpk(this.value())
  }

  /** Deserializes a hexadecimal string into a `DoublePublicKey` instance.
   * @param hex - The hexadecimal string to convert.
   * @returns The `DoublePublicKey` instance represented by the input string.
   */
  static deserialize(
    this: new (obj: any) => DoublePublicKey,
    hex: string
  ): DoublePublicKey {
    return DoublePublicKey._deserialize(hex, deserializeDpk)
  }

  /** Encodes the DoublePublicKey as a bech32m address string.
   * @returns The encoded address string.
   */
  encode(): string {
    const rv = encodeAddress(this.value(), AddressEncoding.Bech32M)
    if (rv.result !== 0) {
      throw new Error(`Failed to encode address: ${rv.result}`)
    }
    // rv.value is a string pointer in WASM, or a string in native
    // The encodeAddress function returns the string directly
    return rv.value as unknown as string
  }

  /** Decodes a bech32/bech32m address string into a DoublePublicKey.
   * @param encoded - The encoded address string.
   * @returns The DoublePublicKey instance.
   */
  static decode(encoded: string): DoublePublicKey {
    const rv = decodeAddress(encoded)
    if (rv.result !== 0) {
      throw new Error(`Failed to decode address: ${rv.result}`)
    }
    return DoublePublicKey.fromObj(rv.value)
  }

  /** Checks if this DoublePublicKey equals another.
   * @param other - The other DoublePublicKey to compare.
   * @returns true if they are equal, false otherwise.
   */
  equals(other: DoublePublicKey): boolean {
    return this.serialize() === other.serialize()
  }
}

File: ffi/rust/src/keys/double_public_key.rs
use crate::{
  blsct_obj::{self, BlsctObj},
  blsct_serde::BlsctSerde,
  ffi::{
    BlsctDoublePubKey,
    BlsctRetVal,
    BlsctSubAddr,
    deserialize_dpk,
    dpk_to_sub_addr,
    gen_double_pub_key,
    gen_dpk_with_keys_acct_addr,
    serialize_dpk,
  },
  sub_addr::SubAddr,
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_value,
  },
  keys::{
    child_key_desc::tx_key_desc::view_key::ViewKey,
    public_key::PublicKey,
  },
};
use std::ffi::c_char;
use serde::{Deserialize, Serialize};

#[derive(PartialEq, Eq, Debug, Deserialize, Serialize)]
pub struct DoublePublicKey {
  obj: BlsctObj<DoublePublicKey, BlsctDoublePubKey>,
}

impl_from_retval!(DoublePublicKey);
impl_display!(DoublePublicKey);
impl_clone!(DoublePublicKey);

impl From<BlsctObj<DoublePublicKey, BlsctDoublePubKey>> for DoublePublicKey {
  fn from(obj: BlsctObj<DoublePublicKey, BlsctDoublePubKey>) -> DoublePublicKey {
    DoublePublicKey { obj }
  }
}

impl From<DoublePublicKey> for SubAddr {
  fn from(dpk: DoublePublicKey) -> SubAddr {
    let rv = unsafe { dpk_to_sub_addr(dpk.value()) };
    BlsctObj::<SubAddr, BlsctSubAddr>::from_retval(rv).unwrap().into()
  }
}

impl DoublePublicKey {
  impl_value!(BlsctDoublePubKey);

  pub fn from_public_keys<'a>(
    pk1: &PublicKey,
    pk2: &PublicKey,
  ) -> Result<Self, blsct_obj::Error<'a>> {
    let rv = unsafe {
      gen_double_pub_key(pk1.value(), pk2.value())
    };
    let obj = BlsctObj::from_retval(rv)?;
    Ok(obj.into())
  }

  pub fn from_keys_acct_addr(
    view_key: &ViewKey,
    spending_pub_key: &PublicKey,
    account: i64,
    address: u64,
  ) -> Self {
    let dpk = unsafe {
      gen_dpk_with_keys_acct_addr(
        view_key.value(),
        spending_pub_key.value(),
        account,
        address,
      )
    };
    let obj = BlsctObj::from_c_obj(dpk);
    obj.into()
  }

  pub fn random<'a>() -> Result<Self, blsct_obj::Error<'a>> { 
    let pk1 = PublicKey::random()?;
    let pk2 = PublicKey::random()?;
    Self::from_public_keys(&pk1, &pk2)
  }
}

impl BlsctSerde for DoublePublicKey {
  unsafe fn serialize(ptr: *const u8, _: usize) -> *const i8 {
    serialize_dpk(ptr as *const BlsctDoublePubKey)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    deserialize_dpk(hex)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    initializer::init,
    keys::child_key::ChildKey,
  };

  #[test]
  fn test_random() {
    init();
    let _: PublicKey = PublicKey::random().unwrap();
  }

  #[test]
  fn test_from_public_keys() {
    init();
    let a = PublicKey::random().unwrap();
    let b = PublicKey::random().unwrap();
    let _: DoublePublicKey = 
      DoublePublicKey::from_public_keys(&a, &b).unwrap();
  }

  #[test]
  fn test_from_keys_acct_addr() {
    init();
    let child_key = ChildKey::random().unwrap();
    let tx_key = child_key.to_tx_key();
    let view_key = tx_key.to_view_key();
    let pub_key = PublicKey::random().unwrap();

    DoublePublicKey::from_keys_acct_addr(
      &view_key,
      &pub_key,
      123,
      456,
    );
  }

  #[test]
  fn test_eq() {
    init();
    let (a, b) = {
      loop {
        let a = DoublePublicKey::random().unwrap();
        let b = DoublePublicKey::random().unwrap();
        if a != b {
          break (a, b);
        }
      }
    };
    assert!(a == a);
    assert!(a != b);
    assert!(b != a);
    assert!(b == b);
  }

  #[test]
  fn test_deser() {
    init();
    let a = DoublePublicKey::random().unwrap();
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<DoublePublicKey>(&hex).unwrap();

    assert_eq!(a, b);
  }
}
